src/libraries/System.Private.CoreLib/src/System/IO/Strategies/WindowsFileStreamStrategy.cs

```diff
@@ -77,9 +78,18 @@ internal WindowsFileStreamStrategy(string path, FileMode mode, FileAccess access
 
         // When the file is locked for writes we can cache file length in memory
         // and avoid subsequent native calls which are expensive.
-        public unsafe sealed override long Length => _share > FileShare.Read ?
-            FileStreamHelpers.GetFileLength(_fileHandle, _path) :
-            _length < 0 ? _length = FileStreamHelpers.GetFileLength(_fileHandle, _path) : _length;
+        public unsafe sealed override long Length
+        {
+            get
+            {
+                if (_share > FileShare.Read || _exposed)
+                {
+                    return FileStreamHelpers.GetFileLength(_fileHandle, _path);
+                }
+
+                return _length = FileStreamHelpers.GetFileLength(_fileHandle, _path);
```

Comment:
Otherwise, this is not caching.
```suggestion
                return _length < 0 ? _length = FileStreamHelpers.GetFileLength(_fileHandle, _path) : _length;
```